lib/wit.js

```diff
@@ -57,76 +62,89 @@ function Wit(opts) {
         logger.warn('Max steps reached, stopping.');
         return prevContext;
       }
-      if (!json.type) {
+      throwIfNotLastIndex(index, this.sessions[sessionId].index);
+      if (!type) {
         throw new Error('Couldn\'t find type in Wit response');
       }
 
       logger.debug('Context: ' + JSON.stringify(prevContext));
-      logger.debug('Response type: ' + json.type);
+      logger.debug('Response type: ' + type);
 
-      // backwards-cpmpatibility with API version 20160516
-      if (json.type === 'merge') {
-        json.type = 'action';
-        json.action = 'merge';
+      // backwards-compatibility with API version 20160516
+      if (type === 'merge') {
+        type = 'action';
+        action = 'merge';
       }
 
-      if (json.type === 'error') {
+      if (type === 'error') {
         throw new Error('Oops, I don\'t know what to do.');
       }
 
-      if (json.type === 'stop') {
+      if (type === 'stop') {
         return prevContext;
       }
 
       const request = {
         sessionId,
         context: clone(prevContext),
         text: message,
-        entities: json.entities,
+        entities
       };
 
-      if (json.type === 'msg') {
+      if (type === 'msg') {
         throwIfActionMissing(actions, 'send');
         const response = {
-          text: json.msg,
-          quickreplies: json.quickreplies,
+          text: msg,
+          quickreplies,
         };
         return actions.send(request, response).then((ctx) => {
           if (ctx) {
             throw new Error('Cannot update context after \'send\' action');
           }
+          throwIfNotLastIndex(index, this.sessions[sessionId].index);
           return this.converse(sessionId, null, prevContext).then(
-            continueRunActions(sessionId, message, prevContext, i - 1)
+            continueRunActions(sessionId, index, message, prevContext, i - 1)
           );
         });
-      } else if (json.type === 'action') {
-        const action = json.action;
+      } else if (type === 'action') {
         throwIfActionMissing(actions, action);
         return actions[action](request).then((ctx) => {
+          throwIfNotLastIndex(index, this.sessions[sessionId].index);
           const nextContext = ctx || {};
           return this.converse(sessionId, null, nextContext).then(
-            continueRunActions(sessionId, message, nextContext, i - 1)
+            continueRunActions(sessionId, index, message, nextContext, i - 1)
           );
         });
       } else {
-        logger.debug('unknown response type', json);
-        throw new Error('unknown response type ' + json.type);
+        logger.debug('unknown response type ' + type);
+        throw new Error('unknown response type ' + type);
       }
     }
   };
 
   this.runActions = (sessionId, message, context, maxSteps) => {
     if (!actions) throwMustHaveActions();
+
+    let {index = 0, text = ''} = this.sessions[sessionId] || {};
+    ++index;
```

Comment:
minor nit:
instead of mutating like so, would some thing like
```
const newSess = {
  index: index + 1,
  text: (text +  ' ' + message).trim()
};
then use newSess below.

main reasoning is -- this conceptually separates the previous values, with the new values
then use
```